### PR TITLE
feat: Room List Bridge

### PR DESCRIPTION
**Pull request by Dnouv**
## Proposed changes (including videos or screenshots)

This pull request introduces a new capability for apps to list and retrieve rooms across the workspace, with flexible filtering options, and adds a lightweight conversion mode for room data to improve performance. The changes affect the Apps Engine bridge, room conversion logic, and the accessors/interfaces that expose these features to apps, along with corresponding tests and type definitions.

**Room Listing API and Filtering Enhancements:**

* Added a new `getAllRooms` method to the Apps Engine bridge and the `IRoomRead` accessor, allowing apps to retrieve rooms workspace-wide with optional filters such as type, pagination, discussion-only, and team-main flags. [[1]](diffhunk://#diff-4a65dde8500650d6a2f892960990af089241a1ed63dcc76a21e5ff6bb1d66bc3R70-R75) [[2]](diffhunk://#diff-bfcf8082cda326cef000bacb8564419b32a77358ef698aff2fd22a845b2814c0R64-R71) [[3]](diffhunk://#diff-d080ea4ed14ebd8efafe8dd146b1c1484e9e7945eb44280a62497ef885a14608R49-R73) [[4]](diffhunk://#diff-47066f8e0fadf4188f21419aa7c4f6ea833bebfde8a267260b3fcc0f94bf3ca2R1-R7)
* Introduced the `GetRoomsFilter` type to specify filtering options for room listing, and enforced validation for filter parameters (e.g., limit and skip) in the accessor implementation. [[1]](diffhunk://#diff-4a65dde8500650d6a2f892960990af089241a1ed63dcc76a21e5ff6bb1d66bc3R18-R26) [[2]](diffhunk://#diff-d080ea4ed14ebd8efafe8dd146b1c1484e9e7945eb44280a62497ef885a14608R49-R73)

**Room Conversion Optimization:**

* Enhanced the room converter (`AppRoomsConverter`) with a new `lightweight` option, enabling conversion of room data without expensive lookups for improved performance when listing many rooms. [[1]](diffhunk://#diff-cd49fd1f2004ffe563d1e85e3824bfb7d58b377e9d5584806ee2d94a5c81ed01L179-R188) [[2]](diffhunk://#diff-cd49fd1f2004ffe563d1e85e3824bfb7d58b377e9d5584806ee2d94a5c81ed01R345-R405) [[3]](diffhunk://#diff-79fb15faa12418115509032633c3c94e7abc7d2bf89e45bdd5acf4e10959461eR5-R14)

**Testing and Type Updates:**

* Updated test fixtures and test data bridges to support the new room listing API and validate filter parameter handling. [[1]](diffhunk://#diff-8e7bb6356297b9660e21086e60ce738836ef18a9f3c5ec73d3d2d9d63e9220a4R58-R60) [[2]](diffhunk://#diff-8e7bb6356297b9660e21086e60ce738836ef18a9f3c5ec73d3d2d9d63e9220a4R79-R86) [[3]](diffhunk://#diff-8e7bb6356297b9660e21086e60ce738836ef18a9f3c5ec73d3d2d9d63e9220a4R99-R100) [[4]](diffhunk://#diff-014c02ade9007c990ef2d15e6c4984b5992812a382bb9f3db061f1044cda16e8R36-R39)
* Extended relevant type definitions and interfaces (`IRoomRead`, `IAppRoomsConverter`) to support new parameters and conversion options. [[1]](diffhunk://#diff-bfcf8082cda326cef000bacb8564419b32a77358ef698aff2fd22a845b2814c0L1-R1) [[2]](diffhunk://#diff-79fb15faa12418115509032633c3c94e7abc7d2bf89e45bdd5acf4e10959461eR5-R14)

These changes collectively enable apps to efficiently list and filter rooms across the workspace, with improved performance and robust API contracts.

## Steps to test or reproduce

<details>
  <summary><strong>Room filter logging (click to expand) 🍎</strong></summary>

  <p>This function applies several room filters (channels, groups, DMs, discussions, pagination, etc.) and logs the matched rooms.</p>

  ```ts
    async executePostMessageSent(
        message: IMessage,
        read: IRead,
        http: IHttp,
        persistence: IPersistence,
        modify: IModify
    ): Promise<void> {
        this.getLogger().log("Message Sent");

        const filters = [
            { label: "All rooms (default)", filters: {}, options: {} },
            { label: "Channels only", filters: { types: [RoomType.CHANNEL] } },
            {
                label: "Private groups",
                filters: { types: [RoomType.PRIVATE_GROUP] },
            },
            { label: "DMs", filters: { types: [RoomType.DIRECT_MESSAGE] } },
            { label: "Discussions only", filters: { discussions: true } },
            {
                label: "Include discussions with channels",
                filters: {
                    types: [RoomType.CHANNEL],
                    discussions: true,
                },
            },
            {
                label: "Team mains (channels)",
                filters: { types: [RoomType.CHANNEL], teams: true },
            },
            { label: "First 5 rooms (paged)", options: { limit: 5, skip: 0 } },
            { label: "Next 5 rooms (paged)", options: { limit: 5, skip: 5 } },
        ];

        for (const { label, filters, options } of filters) {
            const rooms = await read.getRoomReader().getAllRooms(filters, options);
            console.log(`\n${label}: ${rooms.length} rooms`);
            console.log(
                rooms
                    .map((r) => r.displayName || r.slugifiedName || r.id)
                    .join(", ")
            );
        }
    }

```
</details>

## Further comments

[RCAI6-58](https://rocketchat.atlassian.net/browse/RCAI6-58)

[RCAI6-58]: https://rocketchat.atlassian.net/browse/RCAI6-58?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Apps can list multiple rooms with filters, pagination, discussion/team options and receive a new lightweight "raw" room representation including team, parent-room and visitor activity metadata.
  * Added permission to allow apps to view all rooms; translation text included.

* **Tests**
  * Added tests for multi-room reads, limit/skip validation and edge cases.

* **Chores**
  * Bumped minor versions and added model support/indexing for team-related queries.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: packages/apps-engine/src/definition/rooms/RoomType.ts [hunk 1]
```diff
@@ -3,4 +3,6 @@ export enum RoomType {
 	PRIVATE_GROUP = 'p',
 	DIRECT_MESSAGE = 'd',
 	LIVE_CHAT = 'l',
+	DISCUSSION = 'discussion',
+	TEAM = 'team',
```
Comment (d-gubert) [inline packages/apps-engine/src/definition/rooms/RoomType.ts]: This room types do not seem to exist for rooms in the core. Please check the definitions in core-types

File: packages/apps-engine/src/server/accessors/RoomRead.ts [hunk 2]
```diff
@@ -46,6 +46,27 @@ export class RoomRead implements IRoomRead {
 		return this.roomBridge.doGetMembers(roomId, this.appId);
 	}
 
+	public getAllRooms(options: Partial<GetRoomsOptions> = {}): Promise<Array<IRoomRaw>> {
```
Comment (d-gubert) [inline packages/apps-engine/src/server/accessors/RoomRead.ts]: Unclear why you've added `Partial` when all properties of the type are already optional
```suggestion
	public getAllRooms(options: GetRoomsOptions = {}): Promise<Array<IRoomRaw>> {
```

File: apps/meteor/app/apps/server/bridges/rooms.ts [hunk 3]
```diff
@@ -314,4 +344,79 @@ export class AppRoomBridge extends RoomBridge {
...
+
+			if (!includeDiscussions) {
+				baseCondition.prid = { $exists: false };
+			}
+
+			if (!includeTeams) {
+				baseCondition.teamMain = { $ne: true };
+			}
+
+			conditions.push(baseCondition);
```
[the remark this answers is not in the document]
Comment (d-gubert) [reply apps/meteor/app/apps/server/bridges/rooms.ts]: Room types are incorrect, review does not apply

File: packages/apps-engine/src/definition/rooms/IRoomRaw.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	displaySystemMessages?: boolean;
+	messageCount?: number;
+	createdAt?: Date;
+	updatedAt?: Date;
+	lastModifiedAt?: Date;
+	description?: string;
+	customFields?: { [key: string]: any };
+	parentRoomId?: string;
+	teamId?: string;
+	isTeamMain?: boolean;
```
Comment (d-gubert) [inline packages/apps-engine/src/definition/rooms/IRoomRaw.ts:31]: Do you see any reason for us not to add this to the IRoom interface as well?

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 5]
```diff
@@ -337,6 +379,54 @@ export class AppRoomsConverter {
 		return transformMappedData(originalRoom, map);
 	}
 
+	convertRoomWithoutLookups(room) {
```
Comment (d-gubert) [inline apps/meteor/app/apps/server/converters/rooms.js]: Keep naming consistent with the `convertMessageRaw`

```suggestion
	convertRoomRaw(room) {
```

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 6]
```diff
@@ -337,6 +379,54 @@ export class AppRoomsConverter {
...
+					username: room.u.username,
+					name: room.u.name || room.u.username || 'Unknown',
+				}
+			: undefined;
+
+		const fallbackUser = {
+			_id: 'unknown',
+			username: 'unknown',
+			name: 'Unknown',
+		};
```
Comment (d-gubert) [inline apps/meteor/app/apps/server/converters/rooms.js]: Since this method is converting an existing room to match internal apps-engine types, business rules guarantee the `u` property is present, no need to walk around it like this or have a fallback - this fallback is really not helpful at all, it would be better to send `undefined` to the app anyway. But since we'll always have a `u`, there is no need for it

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 7]
```diff
@@ -337,6 +379,54 @@ export class AppRoomsConverter {
...
+	}
+
+	_resolveRoomType(room) {
+		if (room.teamMain) {
+			return RoomType.TEAM;
+		}
+
+		if (room.prid) {
+			return RoomType.DISCUSSION;
+		}
```
Comment (d-gubert) [inline apps/meteor/app/apps/server/converters/rooms.js]: These room types don't exist in the core IRoom interface, they should not exist in the Apps-Engine either

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 8]
```diff
@@ -176,11 +214,15 @@ export class AppRoomsConverter {
...
 
-	async convertRoom(originalRoom) {
+	async convertRoom(originalRoom, { lightweight = false } = {}) {
 		if (!originalRoom) {
 			return undefined;
 		}
 
+		if (lightweight) {
+			return this.convertRoomWithoutLookups(originalRoom);
+		}
```
Comment (d-gubert) [inline apps/meteor/app/apps/server/converters/rooms.js]: There is no need to modify the original method
```suggestion
	async convertRoom(originalRoom) {
		if (!originalRoom) {
			return undefined;
		}
```

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 9]
```diff
@@ -20,6 +20,44 @@ export class AppRoomsConverter {
 		return this.convertRoom(room);
 	}
 
+	async convertRoomRaw(room) {
```
Comment (d-gubert) [inline apps/meteor/app/apps/server/converters/rooms.js]: I'm confused, considering this method exists, `convertRoomWithoutLookup` doesn't need to exist, correct?

File: .changeset/chatty-dingos-bathe.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+---
+'@rocket.chat/apps-engine': minor
+'@rocket.chat/apps': minor
+'@rocket.chat/meteor': minor
+---
+
+Adds room listing to the Apps Engine bridge to let apps retrieve rooms by type and provide workspace-wide visibility.
```
Comment (d-gubert) [inline .changeset/chatty-dingos-bathe.md]:
```suggestion
Adds a new method to the Apps-Engine that allows apps to retrieve multiple rooms from database
```

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 11]
```diff
@@ -20,6 +20,44 @@ export class AppRoomsConverter {
...
+			return undefined;
+		}
+
+		const creator = room.u
+			? {
+					_id: room.u._id,
+					username: room.u.username,
+					name: room.u.name,
+				}
+			: undefined;
```
[the remark this answers is not in the document]
Comment (d-gubert) [reply apps/meteor/app/apps/server/converters/rooms.js]: This object does not need to conform to the `IUser` interface

File: apps/meteor/app/apps/server/converters/rooms.js (hunk 5 above)
Comment (Dnouv) [reply apps/meteor/app/apps/server/converters/rooms.js]: Done!

File: apps/meteor/app/apps/server/converters/rooms.js (hunk 6 above)
Comment (Dnouv) [reply apps/meteor/app/apps/server/converters/rooms.js]: I forgot to remove the unused method `convertRoomWithoutLookups`.

File: apps/meteor/app/apps/server/converters/rooms.js (hunk 9 above)
Comment (Dnouv) [reply apps/meteor/app/apps/server/converters/rooms.js]: Yes, I missed removing it.

Comment (Dnouv) [reply apps/meteor/app/apps/server/converters/rooms.js]: now removed

File: packages/apps-engine/tests/server/accessors/RoomRead.spec.ts [hunk 12]
```diff
@@ -72,6 +89,14 @@ export class RoomReadAccessorTestFixture {
...
 		const rr = new RoomRead(this.mockRoomBridgeWithRoom, 'testing-app');
 
+		let caught = false;
+		try {
+			await rr.getAllRooms({ limit: 200 });
+		} catch (e) {
+			caught = true;
+		}
+
+		Expect(caught).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (d-gubert) [reply packages/apps-engine/tests/server/accessors/RoomRead.spec.ts]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/RocketChat/Rocket.Chat/pull/37719#discussion_r2614495261)

File: packages/apps-engine/src/definition/rooms/RoomType.ts (hunk 1 above)
Comment (Dnouv) [reply packages/apps-engine/src/definition/rooms/RoomType.ts]: there is no type for teams or discussions only some parameters on IRoom, so I reverted the two types to use the `includeDiscussions` and `includeTeamMain`

File: apps/meteor/app/apps/server/converters/rooms.js [hunk 13]
```diff
@@ -20,6 +20,44 @@ export class AppRoomsConverter {
...
+
+		const type = this._convertTypeToApp(room.t);
+
+		return {
+			id: room._id,
+			slugifiedName: room.name || room.fname || room._id,
```
Comment (d-gubert) [inline apps/meteor/app/apps/server/converters/rooms.js]: This mapping for the `slugifiedName` is incorrect 🤔 Refer to the `convertRoom` method for the correct mapping

File: packages/apps-engine/src/server/bridges/RoomBridge.ts [hunk 14]
```diff
@@ -58,6 +68,12 @@ export abstract class RoomBridge extends BaseBridge {
 		}
 	}
 
+	public async doGetAllRooms(options: GetRoomsOptions = {}, appId: string): Promise<Array<IRoomRaw>> {
+		if (this.hasViewAllRoomsPermission(appId)) {
+			return this.getAllRooms(options, appId);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Dnouv) [reply packages/apps-engine/src/server/bridges/RoomBridge.ts]: Good point. But we don't return an empty array anywhere else.

File: apps/meteor/app/apps/server/bridges/rooms.ts [hunk 15]
```diff
@@ -151,6 +185,37 @@ export class AppRoomBridge extends RoomBridge {
 		return promises as Promise<IUser[]>;
 	}
 
+	protected async getAllRooms(filters: GetRoomsFilters = {}, options: GetRoomsOptions = {}, appId: string): Promise<Array<IRoomRaw>> {
+		this.orch.debugLog(`The App ${appId} is getting all rooms with options`, options);
+
+		const { limit = 100, skip = 0 } = options || {};
```
Comment (ggazzo) [inline apps/meteor/app/apps/server/bridges/rooms.ts]:
```suggestion
		const { limit = 100, skip = 0 } = options;
```

File: packages/apps-engine/src/server/bridges/RoomBridge.ts [hunk 16]
```diff
@@ -15,6 +15,35 @@ export type GetMessagesOptions = {
...
+export type GetRoomsFilters = {
+	/**
+	 * When specified, only rooms matching the provided types will be returned.
+	 */
+	types?: Array<RoomType>;
+	/**
+	 * Filter to include or exclude discussion rooms.
+	 *
+	 * When true, discussions are included in the result set.
+	 * When false, discussion rooms are excluded from the result set.
```
Comment (d-gubert) [inline packages/apps-engine/src/server/bridges/RoomBridge.ts]:
```suggestion
	 * Filter to include or exclude discussion rooms. 
	 * 
	 * When undefined (default), discussions are included in the result set.
	 *
	 * When true, ONLY discussions are included in the result set (remove non-discussions).
	 * When false, discussion rooms are excluded from the result set.
```

File: packages/apps-engine/src/server/bridges/RoomBridge.ts [hunk 17]
```diff
@@ -15,6 +15,35 @@ export type GetMessagesOptions = {
...
+	 */
+	discussions?: boolean;
+	/**
+	 * Filter to include or exclude team main rooms.
+	 *
+	 * When true, team main rooms are included in the result set.
+	 * When false, team main rooms are excluded from the result set.
```
Comment (d-gubert) [inline packages/apps-engine/src/server/bridges/RoomBridge.ts:41]:
```suggestion
	 * Filter to include or exclude team main rooms.
	 *
	 * When undefined (default), team main rooms are included in the result set.
	 * 
	 * When true, ONLY team main rooms are included in the result set (remove non-teams).
	 * When false, team main rooms are excluded from the result set.
```
